source/rust_verify/example/syntax_attr.rs

```diff
@@ -0,0 +1,289 @@
+        y < 100,
+)]
+fn test_my_funs(x: u32, y: u32)
+{
+    // my_proof_fun(x, y); // not allowed in exec code
+    // let u = my_spec_fun(x, y); // not allowed exec code
+    proof! {
+        let u = my_spec_fun(x as int, y as int);  // allowed in proof code
+        my_proof_fun(u / 2, y as int);  // allowed in proof code
+    }
+}
+
+verus! {
+
+/// spec functions with pub or pub(...) must specify whether the body of the function
+/// should also be made publicly visible (open function) or not visible (closed function).
+pub open spec fn my_pub_spec_fun1(x: int, y: int) -> int {
+    // function and body visible to all
+    x / 2 + y / 2
+}
+
+/* TODO
+pub open(crate) spec fn my_pub_spec_fun2(x: u32, y: u32) -> u32 {
+    // function visible to all, body visible to crate
+    x / 2 + y / 2
+}
+*/
+
+// TODO(main_new) pub(crate) is not being handled correctly
+// pub(crate) open spec fn my_pub_spec_fun3(x: int, y: int) -> int {
+//     // function and body visible to crate
+//     x / 2 + y / 2
+// }
+pub closed spec fn my_pub_spec_fun4(x: int, y: int) -> int {
+    // function visible to all, body visible to module
+    x / 2 + y / 2
+}
+
+pub(crate) closed spec fn my_pub_spec_fun5(x: int, y: int) -> int {
+    // function visible to crate, body visible to module
+    x / 2 + y / 2
+}
+
+} // verus!
+
+/// Recursive functions must have decreases clauses so that Verus can verify that the functions
+/// terminate.
+#[verus_spec(
+    requires
+        0 < x < 100,
+        y < 100 - x,
+    decreases x,
+)]
+fn test_rec(x: u64, y: u64)
+{
+    if x > 1 {
+        test_rec(x - 1, y + 1);
+    }
+}
+
+verus! {
+
+/// Multiple decreases clauses are ordered lexicographically, so that later clauses may
+/// increase when earlier clauses decrease.
+spec fn test_rec2(x: int, y: int) -> int
+    decreases x, y,
+{
+    if y > 0 {
+        1 + test_rec2(x, y - 1)
+    } else if x > 0 {
+        2 + test_rec2(x - 1, 100)
+    } else {
+        3
+    }
+}
+
+/// Decreases and recommends may specify additional clauses:
+///   - decreases .. "when" restricts the function definition to a condition
+///     that makes the function terminate
+///   - decreases .. "via" specifies a proof function that proves the termination
+///   - recommends .. "when" specifies a proof function that proves the
+///     recommendations of the functions invoked in the body
+spec fn add0(a: nat, b: nat) -> nat
+    recommends
+        a > 0,
+    via add0_recommends
+{
+    a + b
+}
+
+spec fn dec0(a: int) -> int
+    decreases a,
+    when a > 0
+    via dec0_decreases
+{
+    if a > 0 {
+        dec0(a - 1)
+    } else {
+        0
+    }
+}
+
+#[via_fn]
+proof fn add0_recommends(a: nat, b: nat) {
+    // proof
+}
+
+#[via_fn]
+proof fn dec0_decreases(a: int) {
+    // proof
+}
+
+} // verus!
+
+/// variables may be exec, tracked, or ghost
+///   - exec: compiled
+///   - tracked: erased before compilation, checked for lifetimes (advanced feature, discussed later)
```

Comment:
The example below does not show how to use tracked/ghost ? Did I miss some changes to support tracked/ghost variable in verus_spec?